### PR TITLE
Improve `save_to_folder` docstring

### PR DESCRIPTION
**Pull request by h-mayorquin**
The old one is confusing and has references to `extractor_name` which can be confusing as some classess do have that attribute (at least for now).

### CODE REVIEW

File: src/spikeinterface/core/base.py [hunk 1]
```diff
@@ -846,43 +846,63 @@ def save_to_memory(self, sharedmem=True, **save_kwargs) -> "BaseExtractor":
...
-          * extracting traces by calling get_trace() method in chunks
-          * saving data into file (memmap with BinaryRecordingExtractor)
-          * dumping to json/pickle the original extractor for provenance
-          * dumping to json/pickle the cached extractor (memmap with BinaryRecordingExtractor)
+        This method extracts trace data, saves it to a file (using a memory-mapped format
+        with BinaryRecordingExtractor), and stores both the original extractor's provenance
+        and the cached extractor's metadata in JSON format.
 
-        This replaces the use of the old CacheRecordingExtractor and CacheSortingExtractor.
+        The folder final location and name can be specified in a couple of ways ways:
```
Comment (zm711) [inline src/spikeinterface/core/base.py:864]:
```suggestion
        The folder's final location and name can be specified in a couple of ways:
```

File: src/spikeinterface/core/base.py [hunk 2]
```diff
@@ -846,43 +846,63 @@ def save_to_memory(self, sharedmem=True, **save_kwargs) -> "BaseExtractor":
         return cached
 
     # TODO rename to saveto_binary_folder
-    def save_to_folder(self, name=None, folder=None, overwrite=False, verbose=True, **save_kwargs):
+    def save_to_folder(
+        self,
+        name: str = None,
```
Comment (zm711) [inline src/spikeinterface/core/base.py:851]:
```suggestion
        name: str | Path | None = None,
```

Or you don't like adding None as a potential type hint? And below you say Path so it needs Path at minimum.

File: src/spikeinterface/core/base.py [hunk 3]
```diff
@@ -846,43 +846,63 @@ def save_to_memory(self, sharedmem=True, **save_kwargs) -> "BaseExtractor":
...
+        Save the extractor and its data to a folder.
 
-        The save consist of:
-          * extracting traces by calling get_trace() method in chunks
-          * saving data into file (memmap with BinaryRecordingExtractor)
-          * dumping to json/pickle the original extractor for provenance
-          * dumping to json/pickle the cached extractor (memmap with BinaryRecordingExtractor)
+        This method extracts trace data, saves it to a file (using a memory-mapped format
+        with BinaryRecordingExtractor), and stores both the original extractor's provenance
+        and the cached extractor's metadata in JSON format.
```
Comment (zm711) [inline src/spikeinterface/core/base.py:862]:
```suggestion
        and the cached extractor's metadata in JSON format.
```

Is pickle no longer possible at all?

File: src/spikeinterface/core/base.py [hunk 4]
```diff
@@ -846,43 +846,63 @@ def save_to_memory(self, sharedmem=True, **save_kwargs) -> "BaseExtractor":
...
-        Save extractor to folder.
+        Save the extractor and its data to a folder.
 
-        The save consist of:
-          * extracting traces by calling get_trace() method in chunks
-          * saving data into file (memmap with BinaryRecordingExtractor)
-          * dumping to json/pickle the original extractor for provenance
-          * dumping to json/pickle the cached extractor (memmap with BinaryRecordingExtractor)
+        This method extracts trace data, saves it to a file (using a memory-mapped format
+        with BinaryRecordingExtractor), and stores both the original extractor's provenance
```
Comment (zm711) [inline src/spikeinterface/core/base.py:861]:
```suggestion
        This method extracts trace data, saves it to a file (using the memory-mapped
        BinaryRecordingExtractor format), and stores both the original extractor's provenance
```
would probably be the slightly more standard short way to write this.

File: src/spikeinterface/core/base.py (hunk 3 above)
Comment (zm711) [reply src/spikeinterface/core/base.py:862]: Didn't mean this one as a suggestion. Oops!

**Review by zm711 — COMMENTED** [review]
A few comments + typo fixes :)

File: src/spikeinterface/core/base.py (hunk 2 above)
Comment (h-mayorquin) [reply src/spikeinterface/core/base.py:851]: I think that's good. But I don't think name can take a Path. Am I wrong?

Side comment, talking about pet peeves this is one of mine. I actually prefer Optional[str] instead of str | None.  But  I think this is inconsequential and I have been following your standard in this one.

File: src/spikeinterface/core/base.py (hunk 4 above)
Comment (h-mayorquin) [reply src/spikeinterface/core/base.py:861]: Thanks, once we decide on the Path name I will batch all of these suggestions.

File: src/spikeinterface/core/base.py (hunk 3 above)
Comment (h-mayorquin) [reply src/spikeinterface/core/base.py:862]: Not for this function I think. Looking a the body of the function it does not seem like it is an option.

File: src/spikeinterface/core/base.py (hunk 2 above)
Comment (zm711) [reply src/spikeinterface/core/base.py:851]: So my comment was originally because the docstring below said str or Path. But looking at the code it has to be a path because of this line

```python
folder = cache_folder / name
```
`/` can't be used with string. Or we need to do name = Path(name) to make sure it is a path first.

Comment (zm711) [reply src/spikeinterface/core/base.py:851]: > Side comment, talking about pet peeves this is one of mine. I actually prefer Optional[str] instead of str | None. But I think this is inconsequential and I have been following your standard in this one.

I'm okay with Optional[str], but I don't like when someone does Optional[list[float|int']] because it gets too nested. I don't like the numpy default of doing argument, optional to indicate that a default argument is supplied.

File: src/spikeinterface/core/base.py [hunk 5]
```diff
@@ -846,43 +846,63 @@ def save_to_memory(self, sharedmem=True, **save_kwargs) -> "BaseExtractor":
...
         ----------
-        name: None str or Path
-            Name of the subfolder in get_global_tmp_folder()
-            If "name" is given, "folder" must be None.
-        folder: None str or Path
-            Name of the folder.
-            If "folder" is given, "name" must be None.
-        overwrite: bool, default: False
-            If True, the folder is removed if it already exists
+        name : str or Path, optional
```
Comment (zm711) [inline src/spikeinterface/core/base.py:884]:
```suggestion
        name : str or Path, optional
```

currently this is wrong, because only a path would work.

And this type of optional is my pet peeve :)

File: src/spikeinterface/core/base.py (hunk 3 above)
Comment (zm711) [reply src/spikeinterface/core/base.py:862]: Cool sorry I didn't look at the function I was just looking at the diff! We can skip this one.

File: src/spikeinterface/core/base.py (hunk 2 above)
Comment (h-mayorquin) [reply src/spikeinterface/core/base.py:851]: > / can't be used with string. Or we need to do name = Path(name) to make sure it is a path first.

You totally can:

```python
In [1]: from pathlib import Path

In [2]: Path.cwd() / "here"
Out[2]: PosixPath('/home/heberto/development/neuroconv/here')
```

Concerning the typing discussion, this is just a small preference. I would rather keep doing it consistently with what you have been doing : )

File: src/spikeinterface/core/base.py (hunk 3 above)
Comment (h-mayorquin) [reply src/spikeinterface/core/base.py:862]: Good.

File: src/spikeinterface/core/base.py (hunk 2 above)
Comment (zm711) [reply src/spikeinterface/core/base.py:851]: Sorry you're right. I wasn't thinking that folder could be a string. But you can also do:
```
first_path = Path('.')
second_path = Path('folder')
full_path = first_path / second_path
```

So it can be a string or Path.

Comment (h-mayorquin) [reply src/spikeinterface/core/base.py:851]: Oh fair enough. 
I guess that you are right that it **can** be a path and that would work. I don't think this is what it was intended and I would rather not give a type hint for this.

Comment (zm711) [reply src/spikeinterface/core/base.py:851]: Then I would delete it from the docstring as well.

Comment (h-mayorquin) [reply src/spikeinterface/core/base.py:851]: Oh, yes, sure! 

I think that the intention is for you to pass the name of the file. As you mentioned you could pass Path("name") and that would work. But the reason that I think is a bad idea is because if someone is relying on a linter and then they pass Path("C:/something_here/here_something_else/") then that would throw an error but the linter would accept it. 

To be honest, my intuition is that folder is sufficient as it can do the two things. But I think that probably at some point it was found convenient to save to the tmp folder but still name it.

File: src/spikeinterface/core/base.py (hunk 5 above)
Comment (h-mayorquin) [reply src/spikeinterface/core/base.py:884]: What do you mean? what are the preferences with option and what is your preference?

Comment (zm711) [reply src/spikeinterface/core/base.py]: Sorry so it is confusing, but there are two uses of optional--one from std library meaning `value | None`

```python
from typing import Optional
Optional[float]
```

And one from numpy docstring style, which is 

```python
def function(my_arg=1)
 """ my_arg: int, optional
```

notice the difference between Optional (from typing) and optional (lowercase numpy docstring convention)

The difference being Optional says an argument is actually optional (for example you choose one between folder or name so you have the option to pick). optional means that a default is already supplied to the argument of the function so you can optionally change the default value. 

I don't like the numpy style because I prefer to be more explicit in the docstring so instead I prefer
```
my_arg: int, default =1 
```

instead of, 
```
my_arg: int, optional
```

So I've worked toward removing all mentions of argument, default_value, optional from the library.
If someone prefers Optional[x] to mean a default of None that doesn't really bother me.

File: src/spikeinterface/core/base.py (hunk 2 above)
Comment (h-mayorquin) [reply src/spikeinterface/core/base.py]: Should be done.

File: src/spikeinterface/core/base.py (hunk 5 above)
Comment (zm711) [reply src/spikeinterface/core/base.py]: technically Optional[float], optional would be an appropriate numpy docstring because a default is supplied in the arguments of the function. Which I don't like.

Comment (h-mayorquin) [reply src/spikeinterface/core/base.py]: I don't think that you are right about the meaning of `typing.Optional`:

https://docs.python.org/3/library/typing.html#typing.Optional

> Note that this is not the same concept as an optional argument, which is one that has a default. An optional argument with a default does not require the Optional qualifier on its type annotation just because it is optional. For example:

> def foo(arg: int = 0) -> None:
    ...
> On the other hand, if an explicit value of None is allowed, the use of Optional is appropriate, whether the argument is optional or not. For example:

Optional means that it can be None.

Comment (h-mayorquin) [reply src/spikeinterface/core/base.py]: But if anything that makes the need for your disctinction stronger,  if there is a default value, you should say what the default values is.

Comment (h-mayorquin) [reply src/spikeinterface/core/base.py]: I am getting you right if I think that the current docstring is working well with your preference?

File: src/spikeinterface/core/base.py [hunk 6]
```diff
@@ -846,43 +846,63 @@ def save_to_memory(self, sharedmem=True, **save_kwargs) -> "BaseExtractor":
...
-            If "name" is given, "folder" must be None.
-        folder: None str or Path
-            Name of the folder.
-            If "folder" is given, "name" must be None.
-        overwrite: bool, default: False
-            If True, the folder is removed if it already exists
+        name : str , optional
+            The name of the subfolder within the global temporary folder. If `folder`
+            is provided, this argument must be None.
+        folder : str or Path, optional
```
Comment (zm711) [inline src/spikeinterface/core/base.py:887]: folder : Optional[str | Path], default: None

or

folder: str | Path | None , default: None

would be my two preferences. Using `,optional` isn't the `Optional` I'm okay with.

sorry this is tricky. I think my own confusion came from the fact the way I was trying to describe Optional how I thought it would work:

I never expected a function to do `(argument: Optional[str]= 'home_folder')` instead of `(argument: Optional[str] = None)`. But technically you use Optional for both. It just seems like poor design to accept None if it isn't the default argument for me.

Basically numpy optional is to indicated required vs optional arguments, which to me already seems clear based on the function parameters.

```
func(x, y=1)
"""x: float
     y, optional
"""
```

Comment (h-mayorquin) [reply src/spikeinterface/core/base.py:887]:
```
I never expected a function to do (argument: Optional[str]= 'home_folder') instead of (argument: Optional[str] = None). But technically you use Optional for both. It just seems like poor design to accept None if it isn't the default argument for me.
```

I agree with this, I would not like to use optional in that case.

I am more confused about numpy convention, can you link it?
My own preference would be to use default=value if there is a default that is not None  and optional to mean that you don't have to specify it.

Comment (zm711) [reply src/spikeinterface/core/base.py:887]: https://numpydoc.readthedocs.io/en/latest/format.html

Here's their style guide. They don't try to force you to use optional, but if you use vscode numpydoc style then it autofills in the optional. So it is a recommendation that I think the auto-docstring complete takes too far.

Comment (zm711) [reply src/spikeinterface/core/base.py:887]: So I shouldn't blame numpydoc, but instead I should blame our overreliance on tooling :)

Comment (h-mayorquin) [reply src/spikeinterface/core/base.py:887]: OK, so numpy doc optional is more expansive than the one in typing.

So, what do you think of using the following option:
* optional only, when is optional in the typing sense (argument=None)
* default, when there is a default value that is not None.

It seems consistent with their docs as they say
> If the default value would not be used as a value, optional is preferred. These are all equivalent:

So, the docstring here is like this:
values that are optional in the typing sense are marked as optional
values that have a default are marksed as default.

What do you think?

File: src/spikeinterface/core/base.py [hunk 7]
```diff
@@ -846,43 +846,63 @@ def save_to_memory(self, sharedmem=True, **save_kwargs) -> "BaseExtractor":
...
-        Save extractor to folder.
+        Save the extractor and its data to a folder.
 
-        The save consist of:
-          * extracting traces by calling get_trace() method in chunks
-          * saving data into file (memmap with BinaryRecordingExtractor)
-          * dumping to json/pickle the original extractor for provenance
-          * dumping to json/pickle the cached extractor (memmap with BinaryRecordingExtractor)
+        This method extracts trace data, saves it to a file (using the memory-mapped
+        BinaryRecordingExtractor format), and stores both the original extractor's provenance
```
Comment (alejoe91) [inline src/spikeinterface/core/base.py:861]: This can be used also to save a `Sorting` object!

File: src/spikeinterface/core/base.py [hunk 8]
```diff
@@ -846,43 +846,63 @@ def save_to_memory(self, sharedmem=True, **save_kwargs) -> "BaseExtractor":
...
+        verbose : bool, default: True
+            If True, print information about the cache folder being used.
+        **save_kwargs
+            Additional keyword arguments to be passed to the underlying save method.
 
         Returns
         -------
-        cached: saved copy of the extractor.
+        cached_extractor
+            A saved copy of the extractor in the specified format (e.g., BinaryRecordingExtractor).
```
Comment (alejoe91) [inline src/spikeinterface/core/base.py:900]: Actually this function returns a `BinaryFolderRecording`, but it's different for the `Sorting` object

File: src/spikeinterface/core/base.py (hunk 7 above)
Comment (alejoe91) [reply src/spikeinterface/core/base.py:861]: So it can be a:
- `BinaryFolderRecording` for the recording
- `NumpySortingFolder` for the sorting

Comment (h-mayorquin) [reply src/spikeinterface/core/base.py:861]: Yeah, sure.

File: src/spikeinterface/core/base.py (hunk 8 above)
Comment (h-mayorquin) [reply src/spikeinterface/core/base.py:900]: Yep. I rather make no reference to the type though. I don't know if they are helpful.

File: src/spikeinterface/core/base.py (hunk 7 above)
Comment (h-mayorquin) [reply src/spikeinterface/core/base.py:861]: Actually, I would rather remove mentioned of the specific type. I don't think those details are useful.

File: src/spikeinterface/core/base.py (hunk 6 above)
Comment (h-mayorquin) [reply src/spikeinterface/core/base.py:887]: Can I mark this as resolved, @zm711? I moved the discussion to https://github.com/SpikeInterface/SpikeInterface-Hackathon-Edinburgh-May24/issues/29

File: src/spikeinterface/core/base.py (hunk 7 above)
Comment (h-mayorquin) [reply src/spikeinterface/core/base.py]: Removed.

File: src/spikeinterface/core/base.py (hunk 8 above)
Comment (h-mayorquin) [reply src/spikeinterface/core/base.py]: Removed.

File: src/spikeinterface/core/base.py (hunk 6 above)
Comment (h-mayorquin) [reply src/spikeinterface/core/base.py:887]: @zm711 ?

Comment (h-mayorquin) [reply src/spikeinterface/core/base.py:887]: Thanks, man.

**Comment by h-mayorquin** [comment]
This is ready.
